runtime/trampolines.m

```diff
@@ -201,7 +201,17 @@
 					}
 				}
 			} else {
-				xamarin_assertion_message ("Don't know how to marshal a return value of type '%s.%s'. Please file a bug with a test case at https://github.com/xamarin/xamarin-macios/issues/new\n", mono_class_get_namespace (r_klass), mono_class_get_name (r_klass)); 
+				char *fullname = xamarin_class_get_full_name (r_klass, exception_gchandle);
```

Comment:
last little bit - since `fullname` is only used in the error and @rolfbjarne implies that it's more costly than the `xamarin_is_class_` tests, how about moving the allocation and deallocation into the `else` clause so that the price only gets paid in the failing case.

Reply:
I reverted the assertion to the original code as it did not need (extra) allocation, inside the caller, and still produced the same output.